lib/AST/ASTImporter.cpp

```diff
@@ -2339,6 +2360,14 @@ bool ASTNodeImporter::ImportTemplateInformation(FunctionDecl *FromFD,
 
 Decl *ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
 
+  llvm::SmallVector<Decl*, 2> Redecls = getCanonicalForwardRedeclChain(D);
```

Comment:
I was wondering if only importing two decl, the canonical and the one with the definition is sufficient? 
In case it is sufficient, I think it would make the code shorter and easier.

Reply:
There are cases when this simply would not work. E.g.:
```
struct X { friend void f(){} };
void f();    // this makes f available to normal
             // lookup, otherwise it is hidden
```

Here the canonical have the definition, and the existence of the second decl changes the semantics of the lookup!

Reply:
Does this patch fix this at all? I mean, if we already have a definition available we will never try to import the function.

Reply:
Yes, this patch fix it. See this test: https://github.com/Ericsson/clang/pull/329/files#diff-10501f0b3ce0b0d6c9b9b9a7d46131afR2033
This is working because when any element of a decl chain is imported then we import the whole decl chain as it is in the "from" context.

Reply:
Q1: Why do we need the current order and why the order by `redecls` is not sufficient?
Q2: I wonder if this will be quadratic in the length of the declaration chain? We will import each member of the redecl chain and for each member of the chain we will enumerate all the forward decls before that declaration.

Reply:
A1: We don not need to keep the current order, it is not necessary ... or at least I don't know any cases which might result different semantics if the elements of the chain are in different order. However keeping the original order would preserver the original semantics for sure.
But, during import of each element of the chain we need a unified view of the chain and we cannot get that unified view with `redecls()`. 
`Decl::redecls()` gives a list of decls which always start from the previous decl and the next item is actually the previous item in the order of source locations.  Thus, `Decl::redecls()` gives different lists for the different entries in a given redecl chain.
```
// Some notes on redeclarables:
//
//  - Every redeclarable is on a circular linked list.
//
//  - Every decl has a pointer to the first element of the chain _and_ a
//    DeclLink that may point to one of 3 possible states:
//      - the "previous" (temporal) element in the chain
//      - the "latest" (temporal) element in the chain
//      - the an "uninitialized-latest" value (when newly-constructed)
//
//  - The first element is also often called the canonical element. Every
//    element has a pointer to it so that "getCanonical" can be fast.
//
//  - Most links in the chain point to previous, except the link out of
//    the first; it points to latest.
//
//  - Elements are called "first", "previous", "latest" or
//    "most-recent" when referring to temporal order: order of addition
//    to the chain.
//
//  - To make matters confusing, the DeclLink type uses the term "next"
//    for its pointer-storage internally (thus functions like
//    NextIsPrevious). It's easiest to just ignore the implementation of
//    DeclLink when making sense of the redeclaration chain.
//
//  - There's also a "definition" link for several types of
//    redeclarable, where only one definition should exist at any given
//    time (and the defn pointer is stored in the decl's "data" which
//    is copied to every element on the chain when it's changed).
//
//    Here is some ASCII art:
//
//      "first"                                     "latest"
//      "canonical"                                 "most recent"
//      +------------+         first                +--------------+
//      |            | <--------------------------- |              |
//      |            |                              |              |
//      |            |                              |              |
//      |            |       +--------------+       |              |
//      |            | first |              |       |              |
//      |            | <---- |              |       |              |
//      |            |       |              |       |              |
//      | @class A   |  link | @interface A |  link | @class A     |
//      | seen first | <---- | seen second  | <---- | seen third   |
//      |            |       |              |       |              |
//      +------------+       +--------------+       +--------------+
//      | data       | defn  | data         |  defn | data         |
//      |            | ----> |              | <---- |              |
//      +------------+       +--------------+       +--------------+
//        |                     |     ^                  ^
//        |                     |defn |                  |
//        | link                +-----+                  |
//        +-->-------------------------------------------+
```

A2: No, because once a member is imported then all other import calls return early with the already imported decl. Consider the import of `C` from this chain `A<-B<-C`.
```
VisitFunctionDecl(C)
  VisitFunctionDecl(A) // A is imported before any other recursive import to the chain 
    VisitFunctionDecl(B)
       Import(A) // A is already imported, early return
       VisitFunctionDecl(C) 
          Import(A) // A is already imported, early return
          Import(B) // B is already imported, early return
          // C is imported
```
Each VisitFunctionDecl calls `getCanonicalForwardRedeclChain` which is O(n). As a summary this looks as O(n^n). But 99% of the time n <= 2.

Reply:
I see, so the visit calls are not quadratic but the hash table lookups in the already imported map is. But that might lets first see if that is a problem at all.